### PR TITLE
Added a custom Debug Adapter (VSC-250)

### PR DESCRIPTION
**Pull request by an-dr**
Added support of debugging using [DAP](https://microsoft.github.io/debug-adapter-protocol/)-compatible debug adapter

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -0,0 +1,33 @@
+asn1crypto==0.24.0
+astroid==1.6.5
+backports.functools-lru-cache==1.5
+certifi==2018.11.29
+cffi==1.11.5
+chardet==3.0.4
+click
+colorama==0.4.1
+configparser==3.7.1
+cryptography==2.4.2
```
Comment (igrr) [inline requirements.txt]: sorry, can you check which package pulls in `cryptography`?

File: src/test/adapter.test.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    });
+
+    suite("setBreakpoints", () => {
+        test("should stop at breakpoint", () => {
+            const launchArgs = { elf: "path/to/elf"}; // Launch arguments are passed here, if any.
+            const breakpointLine = 2;
+
+            return debugClient.hitBreakpoint(launchArgs, { path: launchArgs.elf, line: breakpointLine});
+        });
+    }); */
```
Comment (igrr) [inline src/test/adapter.test.ts]: not needed => can be removed?

Comment (igrr) [reply src/test/adapter.test.ts]: +1 above

File: templates/fibonacci-app/main/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# Edit following two lines to set component requirements (see docs)
+set(COMPONENT_REQUIRES )
+set(COMPONENT_PRIV_REQUIRES )
+
+set(COMPONENT_SRCS "main.c")
+set(COMPONENT_ADD_INCLUDEDIRS "")
+
+register_component()
```
Comment (igrr) [inline templates/fibonacci-app/main/CMakeLists.txt]: is it possible to use idf_component_register instead?

File: templates/.vscode/settings.json [hunk 4]
```diff
@@ -7,5 +7,11 @@
     "[c]": {
         "editor.quickSuggestions": true
     },
-    "C_Cpp.intelliSenseEngineFallback": "Disabled"
+    "C_Cpp.intelliSenseEngineFallback": "Disabled",
+    
+    "idf.adapterTargetName": "Esp32_S2",
```
Comment (igrr) [inline templates/.vscode/settings.json]: I'm not sure, does this automatically get changed depending on the selected target?

Comment (an-dr) [reply templates/.vscode/settings.json]: it does

**Comment by an-dr** [comment]
@brianignacio5 after rebasing to 0.1.3 I've faced with several issues, need your participation. I will describe them bellow. Thanks!

**Comment by an-dr** [comment]
If there is no `Openocd` in the `Path` - silently do nothing

**Comment by an-dr** [comment]
While debugging getting

Image 001


and after this debugger ceases to reply to step, continue and other buttons

**Comment by an-dr** [comment]
Previosly I used `${config:idf.openOcdBin}`/`${config:idf.openOcdBinWin}` and `${config:idf.openOcdScriptsPath}`/`${config:idf.openOcdScriptsPathWin}` to launching adapter manually from [tasks.json](https://github.com/an-dr/vscode-esp-idf-extension/blob/7544f14b395aefdb04474ba0725fc50d0e739386/templates/.vscode/tasks.json#L120-L162). Could you advice to which parameters it is better to change them now?

**Comment by brianignacio5** [comment]
Hey @an-dr I've just made a PR to your repo with some changes in vscode part.

I took your PR, rebase it to current master in this repo, and add some changes. 

Please take a look in https://github.com/an-dr/vscode-esp-idf-extension/pull/1

I tried to plug a build and flash before debug (thereby this error on debug) which was working on MacOS but is not working for you or me in Ubuntu. Will remove for now and think about solutions for it later.

With regards to `${config:idf.openOcdScriptsPath}/${config:idf.openOcdScriptsPathWin}` we replaced it with `${config:idf.customExtraPaths}` as shown in the `tasks.json` in the mentioned PR. 

Let me know further steps.

**Comment by an-dr** [comment]
@brianignacio5 works now! Testing remained. Can't see if I can do enything to fix this
```
  <testsuite name="initialize" timestamp="2020-02-25T09:26:38" tests="1" file="c:\esp\github\vscode-esp-idf-extension\d                                  ebug-adapter\out\test\suite\adapter.test.js" time="0.0000" failures="1">
    <testcase name="Debug Adapter Tests initialize should return supported features" time="0.0000" classname="should re                                  turn supported features">
      <failure message="Cannot read property &apos;write&apos; of undefined" type="TypeError"><![CDATA[TypeError: Canno                                  t read property 'write' of undefined
        at EspIdfDebugClient.doSend (c:\esp\github\vscode-esp-idf-extension\debug-adapter\node_modules\vscode-debugadap                                  ter-testsupport\lib\protocolClient.js:46:27)
        at c:\esp\github\vscode-esp-idf-extension\debug-adapter\node_modules\vscode-debugadapter-testsupport\lib\protoc                                  olClient.js:24:18
        at new Promise (<anonymous>)
        at EspIdfDebugClient.send (c:\esp\github\vscode-esp-idf-extension\debug-adapter\node_modules\vscode-debugadapte                                  r-testsupport\lib\protocolClient.js:23:16)
        at EspIdfDebugClient.initializeRequest (c:\esp\github\vscode-esp-idf-extension\debug-adapter\out\test\espIdfDeb                                  ugClient.js:90:21)
        at Object.<anonymous> (c:\esp\github\vscode-esp-idf-extension\debug-adapter\out\test\suite\adapter.test.js:48:3                                  1)
        at Generator.next (<anonymous>)
        at c:\esp\github\vscode-esp-idf-extension\debug-adapter\out\test\suite\adapter.test.js:24:71
        at new Promise (<anonymous>)
        at __awaiter (c:\esp\github\vscode-esp-idf-extension\debug-adapter\out\test\suite\adapter.test.js:20:12)                                                 at Context.<anonymous> (c:\esp\github\vscode-esp-idf-extension\debug-adapter\out\test\suite\adapter.test.js:41:                                  56)
        at processImmediate (internal/timers.js:439:21)]]></failure>
    </testcase>
  </testsuite>
```

**Comment by an-dr** [comment]
@brianignacio5 please help with conflicts

**Comment by brianignacio5** [comment]
Please check https://github.com/an-dr/vscode-esp-idf-extension/pull/5

**Comment by an-dr** [comment]
@brianignacio5 bria still exist :(

File: esp_debug_adapter/debug_adapter/debug_backend/Pipfile [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+[dev-packages]
+
+[packages]
+future = ">=0.17"
+psutil = "==5.5.1"
+pygdbmi = "==0.9.0.0"
+typing = ">=3.6.6"
+xmlrunner = "==1.7.7"
+unittest-xml-reporting = "*"
```
Comment (gerekon) [inline esp_debug_adapter/debug_adapter/debug_backend/Pipfile]: @an-dr I think we need to clean up `debug_backend` dependencies. From my point of view we need to leave `pygdbmi` only here.

Comment (gerekon) [reply esp_debug_adapter/debug_adapter/debug_backend/Pipfile]: @brianignacio5 FYI

**Comment by igrr** [comment]
Yay!

**Pull request merged**